lib/trino-filesystem-alluxio/src/main/java/io/trino/filesystem/alluxio/ConsistentHashingCacheHostAddressProvider.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.alluxio;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import com.google.inject.Inject;
+import io.airlift.log.Logger;
+import io.trino.filesystem.cache.CacheHostAddressProvider;
+import io.trino.spi.HostAddress;
+import io.trino.spi.Node;
+import io.trino.spi.NodeManager;
+import jakarta.annotation.PostConstruct;
+import jakarta.annotation.PreDestroy;
+import org.ishugaliy.allgood.consistent.hash.ConsistentHash;
+import org.ishugaliy.allgood.consistent.hash.HashRing;
+import org.ishugaliy.allgood.consistent.hash.hasher.DefaultHasher;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+
+public class ConsistentHashingCacheHostAddressProvider
+        implements CacheHostAddressProvider
+{
+    private static final Logger log = Logger.get(ConsistentHashingCacheHostAddressProvider.class);
+
+    private final NodeManager nodeManager;
+    private final ScheduledExecutorService hashRingUpdater = newSingleThreadScheduledExecutor(daemonThreadsNamed("hash-ring-refresher-%s"));
+
+    private final ConsistentHash<TrinoNode> consistentHashRing = HashRing.<TrinoNode>newBuilder()
+            .hasher(DefaultHasher.METRO_HASH)
+            .build();
+
+    @Inject
+    public ConsistentHashingCacheHostAddressProvider(NodeManager nodeManager)
+    {
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
+    }
+
+    @Override
+    public List<HostAddress> getHosts(String identifier)
+    {
+        return consistentHashRing.locate(identifier)
```

Comment:
`consistentHashRing.locate(identifier)` always returns just a single node, so this method will return just a single HostAddress. Are we intentionally caching each split just on a single node? If there's some hot data, wouldn't too many of the splits be scheduled on that single node where the split is cached, either overwhelming it or increasing the external read ratio? Would it make sense to cache the same split on multiple workers (up to a maximum K number of nodes, i.e. `consistentHashRing.locate(identifier, K)`) to avoid this bottleneck?

Reply:
hey @jkylling just wanted to bump this question up to see what you think. Thanks!

Reply:
Hey! Thanks for the question. We could definitely make the K a configuration option. There is some previous discussion on this topic here: https://github.com/trinodb/trino/pull/18719#discussion_r1318507465